TEvtGen/THepMCParser/THepMCParser.cxx

```diff
@@ -436,154 +439,445 @@ string THepMCParser::ParseGenEvent2TCloneArray(HepMC::GenEvent * genEvent, TClon
+//            AliWarningGeneralStream("THepMCParser") << "Mothers already set to "
+//            << daughter->GetFirstMother()
+//            << ", " << daughter->GetSecondMother()
+//            << ", while trying to set: "
+//            << motherA
+//            << ", " << motherB
+//            << std::endl;
+//          }
+        }
+      }
+      Int_t lastDaughter = arrayID - 1;
+      if (lastDaughter < firstDaughter) {
+        //AliDebugGeneralStream("THepMCParser", 3) << "Vertex with no out particles, should not be possible!" << std::endl;
+      }
+      else {
+        // update mother with daughter interval
+        if (motherA > -1) {
+          if (motherA_p->GetFirstDaughter() > -1) {
+            //AliDebugGeneralStream("THepMCParser", 3) << "Trying to assign new daughters to a particle that already has daughters defined! "
+            //<< "(motherA, id = " << motherA
+            //<< ", event = " << genEvent->event_number()
+            //<< ")" << std::endl;
+          }
+          else {
+            //AliDebugGeneralStream("THepMCParser", 3) << "Setting daughter range for mother A to [" << firstDaughter << ", " << lastDaughter << "]" << std::endl;
+            motherA_p->SetFirstDaughter(firstDaughter);
+            motherA_p->SetLastDaughter(lastDaughter);
+          }
+        }
+//        if (motherB > -1) {
+//          if (motherB_p->GetFirstDaughter() > -1 && motherB > 1) {
+//            AliDebugGeneralStream("THepMCParser", 3) << "Trying to assign new daughters to a particle that already has daughters defined! "
+//            << "(motherB, id = " << motherB
+//            << ", event = " << genEvent->event_number()
+//            << ")" << std::endl;
+//          }
+//          else {
+//            AliDebugGeneralStream("THepMCParser", 3) << "Setting daughter range for mother B to [" << firstDaughter << ", " << lastDaughter << "]" << std::endl;
+//            motherB_p->SetFirstDaughter(firstDaughter);
+//            motherB_p->SetLastDaughter(lastDaughter);
+//          }
+//        }
+      }
+    }
+    if (old_vertex_iterator == vertex_iterator[ibeamPart] || initial_particles == arrayID) {
+      no_advance++;
+      //AliDebugGeneralStream("THepMCParser", 3) << "No advances made in this loop. "
+      //<< "No advance count is " << no_advance << std::endl;
+    }
+    else {
+      no_advance = 0;
+    }
+    if (no_advance > 3) {
+      //AliDebugGeneralStream("THepMCParser", 3) << "No advances for > 3 times. Stopping." << std::endl;
+      break;
+    }
+    if (switchBeamPart) {
+      //AliDebugGeneralStream("THepMCParser", 3) << "Switching to the tree of the other beam particle to look for mother particle." << std::endl;
+      continue;
+    }
+  } while (true);
+
+//  AliDebugGeneralStream("THepMCParser", 3) << "Particles in event: " << nParticlesInHepMC
+//  << ", added: " << arrayID
+//  << ", entries: " << array->GetEntries() << "/" << array->GetEntriesFast()
+//  << std::endl;
```

Comment:
TL,DR 

It seems to me the strategy should be 

- Given a vertex 
  - Register all out-going particle 
    - Check in mapping from GenParticle* to array index
      - If mapping exist, get particle 
      - otherwise create new particle in clones array from GenParticle and set mapping
   - Set first and last daughter index on incoming particles via mapping from GenParticle* to index 
    - For each outgoing particle - check if they have an end vertex 
      - If they do, do this procedure on that vertex 

This will then recurs the tree (yes, recursive calls). 

What is "special" in this case are particles without a production vertex (presumably beam particles).